### DISCUSSION TITLE
Feature Request: Configurable audio latency

### DISCUSSION THREAD

**Discussion by elyorkhakimov** [ExistentialAudio/BlackHole#185] [Feature Requests]
Hi, first of all, I love this software - simple, effective and very useful!

I am struggling a bit with an issue: keeping the resulting AV stream in sync. I have USB capture cards getting HDMI signal from external camera, and using either Macbook's onboard or even USB microphone. Both audio and video sources can be selected in Zoom, Webex, skype, etc., and the result is - video is a bit delayed (about 8-12 frames at 30fps) when compared to audio track.

This feature request is to provide a configurable sync offset, such that in a BlackHole virtual device, I could add my source microphone, add custom delay, and then select the BlackHole virtual device as a microphone in Zoom.

Thanks!

**Comment by devinroth** [comment]
There is an option to add latency to do the driver. Check out the wiki. Maybe that’s a solution.

**Comment by elyorkhakimov** [comment]
Wow, I appreciate super fast response!
I did checkout the wiki page: https://github.com/ExistentialAudio/BlackHole/wiki/Adjust-Driver-Latency
But it sounds like hard-coding the latency for all devices, inputs/outputs. 

Would it be possible to expose this as a configurable option in the user interface?

**Comment by elyorkhakimov** [comment]
ok, as a temp workaround I've re-built the driver, but I am stuck actually at the configuration in the MIDI setup in UI.

I am trying to figure out a way to select "delayed" audio input as a microphone in Zoom's settings.
How do I create this type of "delayed" audio device?

**Comment by devinroth** [comment]
I don't think I understand how you are stuck. Maybe send some screenshots.

**Comment by elyorkhakimov** [comment]
I've got Multiple-Output device configured, and this takes care of me being able to record sound output from headphones. 
![image](https://user-images.githubusercontent.com/22326769/88120336-8750fd00-cb77-11ea-831d-77a586c6f61f.png)

But in addition to the above, I am trying to solve another problem: I have a video input that shows up as a webcam (Cam Link 4k) but it has a delay of 8-12 frames when compared to my external microphone. In other words, if I select this webcam along with external microphone, the resulting audio+video recording has a perceivable sync issue.

I am trying to solve this by creating a virtual microphone that could take the real external microphone, add a delay of 200-300msec and then use this virtual microphone from Zoom/Skype. How can I do this with BlackHole ?

**Comment by devinroth** [comment]
Download GarageBand.
Set the input to your mic.
Set the output to BlackHole.
Make sure Input Monitoring is enabled.
Add Sample Delay to the track. 
Adjust Sample Delay until everything lines up.

Should solve your problem.

**Comment by julian-weinert** [comment]
I can also recommend _Hosting AU_ for such a small task.
That way you don't need to run a "full" DAW

**Comment by ghost** [comment]
I have a similar issue, setup:
- Nikon DSLR D3300 Camera -> USB HDMI Capture Card -> Mac USB
- Any microphone not onboard the camera (built-in Mac, line-in headphones, bluetooth headphones)

The capture card has its own latency, so when using that as camera feed in Zoom, my video is delayed slightly. So I need to delay my microphone input by a considerable amount of ms before being fed into zoom to match the video feed.

I downloaded the HostingAU recommendation above but it seems like I still need something like VB Cable or Blackhole, since HostingAU allows me to take an input and apply up to 5 AUSampleDelay effects (each has max 4000smps and so one effect isn't enough), but it needs an output to send it to, and zoom mic can only take an input, so VB Cable/Blackhole is this piece in the middle needed to act as a converter of output from Hosting AU/Garageband to input for Zoom

MIC -> Hosting AU (apply delay here) -> BlackHole -> Zoom Input

@julian-weinert did you mean that Hosting AU by itself would allow us to do this direct to Zoom as input? Or did you mean we'd still need BlackHole / VB Cable, and Hosting AU would allow the delay effect only?

Just trying to get as lean as possible before figuring out the exact delay I need

**Comment by contentfree** [comment]
I have a similar set up to @ghost above: Nikon DSLR connected to a Mac via a USB Capture Card (Cam Link 4K) which introduces a video delay of ~200ms. I came to the BlackHole project specifically because it was recommended as a solution for introducing audio delay on the line so that I could sync audio and video for Teams or Zoom calls. Is Blackhole *not* the right solution, @devinroth?

**Reply by devinroth** [reply]
Yeah it can work. You could add a custom delay in the signal flow so things line up.
